README.md
tranSMART 1.2 Install
=====================

This install is based on the install instructions contained in:

https://wiki.transmartfoundation.org/pages/viewpage.action?pageId=6619205

The RHEL 6.5 and 7.0 instructions are incomplete on this page so that install 
diverges somewhat. 








CHANGELOG.md
# CHANGELOG for tranSmart

This file is used to list changes made in each version of tranSmart.

## 0.1.0:

* Initial release of tranSmart

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


test/integration/default/serverspec/git_daemon_spec.rb
require 'serverspec'

# Required by serverspec
set :backend, :exec

describe command('which git') do
  its(:exit_status) { should eq 0 }
end

describe command('which java') do
  its(:exit_status) { should eq 0 }
end

describe command('which psql') do
  its(:exit_status) { should eq 0 }
end

describe command('which make') do
  its(:exit_status) { should eq 0 }
end

describe command('ls -d transmart-data') do
  its(:stdout) { should match(/transmart-data/) }
end

describe command('stat transmart-data/vars') do
  its(:exit_status) { should eq 0 }
end

if os[:family] == 'redhat'
  describe command('ps -ef')  do
    its(:stdout) { is_expected.to include 'postgres' }
  end

  describe command('ps -ef')  do
    its(:stdout) { is_expected.to include 'apache ' }
  end
end

if os[:family] == 'ubuntu'
  describe command('ps -ef') do
    its(:stdout) { is_expected.to include 'postgresql' }
  end

  describe command('ps -ef')  do
    its(:stdout) { is_expected.to include 'apache2' }
  end
end

describe command('ps -ef') do
  its(:stdout) { is_expected.to include 'tomcat7' }
end

describe 'Tomcat Service' do
  it 'is listening on port 8080' do
    expect(port(8080)).to be_listening
  end
end

describe 'SOLR Service' do
  it 'is listening on port 8983' do
    expect(port(8983)).to be_listening
  end
end

describe 'Postges Service' do
  it 'is listening on port 5432' do
    expect(port(5432)).to be_listening
  end
end


recipes/debian.rb
log 'debian started'

###############################################################################
log 'Set tranSMART repository'

apt_repository 'hyve_internal' do
  uri 'http://apt.thehyve.net/internal/'
  distribution 'trusty'
  components ['main']
  keyserver 'keyserver.ubuntu.com'
  key '3375DA21'
end

###############################################################################
log 'Fix debian bug by adding locale file'

cookbook_file 'locale' do
  path '/etc/default/locale'
  action :create
  owner 'root'
  group 'root'
  mode 0755
end

###############################################################################
log 'Install packages'

include_recipe 'apt'

node['unixNonSpecific'].each do |pkg|
  package pkg do
    action :install
  end
end

node['unixDebian'].each do |pkg|
  package pkg do
    action :install
  end
end

###############################################################################

log 'Get tranSMART from Git'

transmart_data = ENV['PWD'] + '/transmart-data'

include_recipe 'git'

git transmart_data  do
  repository 'https://github.com/transmart/transmart-data.git'
  revision 'master'
  action :checkout
end

###############################################################################

log 'Install tranSMART'

# needed for tomcat to run SOLR on right port
cookbook_file 'configGroovy.patch' do
  path "#{Chef::Config[:file_cache_path]}/configGroovy.patch"
  action :create
end

bash 'Install_tranSMART' do
  user 'root'
  cwd transmart_data
  code <<-EOH
    env > enviromentVariablesStart
    export TAR_COMMAND=tar
    php env/vars-ubuntu.php > vars
    source vars
    sudo -u postgres bash -c "source vars; PGSQL_BIN=/usr/bin/ PGDATABASE=template1 make -C ddl/postgres/GLOBAL tablespaces"
    make postgres
    bash -c "source vars; TSUSER_HOME=/usr/share/tomcat7/ make -C config/ install"
    make -C solr/ start &
    echo 'USER=tomcat7' | sudo tee /etc/default/rserve
    service rserve start
    
    service tomcat7 stop
    echo 'JAVA_OPTS="-Xmx4096M -XX:MaxPermSize=1024M"' | sudo tee /usr/share/tomcat7/bin/setenv.sh
    wget -P /var/lib/tomcat7/webapps/ https://ci.transmartfoundation.org/browse/SAND-TRAPP/latest/artifact/shared/transmart.war/transmart.war
    patch -b -N /usr/share/apache-tomcat-7/tomcat/.grails/transmartConfig/Config.groovy  #{Chef::Config[:file_cache_path]}/configGroovy.patch
    chown tomcat7 /usr/share/apache-tomcat-7/tomcat/.grails/transmartConfig/Config.groovy 
    service tomcat7 start
    
    make -C env/ data-integration
    make -C env/ update_etl
    env > enviromentVariablesEnd
    #make -C samples/postgres load_clinical_GSE8581
    #make -C samples/postgres load_ref_annotation_GSE8581
    #make -C samples/postgres load_expression_GSE8581
  EOH
end

log 'debian Finished!'
###############################################################################
###############################################################################


recipes/rhel65.rb
log 'rhel 6.5 started'

###############################################################################
log 'Set tranSMART repository and add postgres RPM'

execute 'yum install -y http://yum.postgresql.org/9.3/redhat/rhel-7-x86_64/pgdg-redhat93-9.3-1.noarch.rpm' do
  returns [0, 1]
end

cookbook_file 'thehyve_internal.repo' do
  path '/etc/yum.repos.d/thehyve_internal.repo'
  action :create
  owner 'root'
  group 'root'
  mode 0755
end

###############################################################################
log 'Install packages'

node['unixRhel'].each do |pkg|
  package pkg do
    action :install
  end
end

node['unixNonSpecific'].each do |pkg|
  package pkg do
    action :install
  end
end

###############################################################################
log 'Configure Postgres'

execute '/sbin/service postgresql-9.3 initdb' do
end

cookbook_file 'pg_hba.conf_trusted' do
  path '/var/lib/pgsql/9.3/data/pg_hba.conf'
  action :create
  owner 'postgres'
  group 'postgres'
  mode 0600
end

###############################################################################
log 'Install and configure Tomcat 7'

bash 'Install_tomcat7' do
  user 'root'
  cwd '/usr/share/'
  code <<-EOH
    wget -O apache-tomcat-7.tar.gz http://archive.apache.org/dist/tomcat/tomcat-7/v#{node['tomcatVersion']}/bin/apache-tomcat-#{node['tomcatVersion']}.tar.gz
    tar xzf apache-tomcat-7.tar.gz
    mv apache-tomcat-#{node['tomcatVersion']} apache-tomcat-7
    groupadd tomcat7
    useradd -g tomcat7 -d /usr/share/apache-tomcat-7/tomcat tomcat7
    chown -Rf tomcat7.tomcat7 /usr/share/apache-tomcat-7/
  EOH
end

cookbook_file 'tomcat7' do
  path '/etc/init.d/tomcat7'
  action :create
  owner 'tomcat7'
  group 'tomcat7'
  mode 0755
end

###############################################################################
log 'Get tranSMART from Git'

transmart_data = ENV['PWD'] + '/transmart-data'

include_recipe 'git'

git transmart_data  do
  repository 'https://github.com/transmart/transmart-data.git'
  revision 'master'
  action :checkout
end

cookbook_file 'vars' do
  path '/home/ec2-user/transmart-data/vars'
  action :create
  owner 'root'
  group 'root'
  mode 0755
end

###############################################################################
log 'Install tranSMART'

# Needed to change for RServe
cookbook_file 'sudoers.patch' do
  path "#{Chef::Config[:file_cache_path]}/sudoers.patch"
  action :create
end

# needed for tomcat to run SOLR on right port
cookbook_file 'configGroovy.patch' do
  path "#{Chef::Config[:file_cache_path]}/configGroovy.patch"
  action :create
end

bash 'Install_tranSMART' do
  user 'root'
  cwd transmart_data
  code <<-EOH
    apachectl start
    env > enviromentVariablesStart
    chgrp -R postgres /home/ec2-user
    chmod g+rx /home/ec2-user
    export TAR_COMMAND=tar
    source vars

    patch -b -N --verbose /etc/sudoers #{Chef::Config[:file_cache_path]}/sudoers.patch

    service postgresql-9.3 start
    sudo -u postgres bash -c "source vars; PGSQL_BIN=/usr/bin/ PGDATABASE=template1 make -C ddl/postgres/GLOBAL tablespaces"
    make postgres

    bash -c "source vars; TSUSER_HOME=/usr/share/apache-tomcat-7/tomcat/ make -C config/ install"
    chown -Rf tomcat7.tomcat7 /usr/share/apache-tomcat-7/
    make -C solr/ start &

    echo 'USER=tomcat7' | sudo tee /etc/default/rserve # not sure this is needed in this version?

    bash -c 'echo "nohup /opt/R/bin/R  CMD Rserve --quiet --vanilla --RS-conf /etc/Rserve.conf&" > /tmp/rserve.sh' && chmod a+rx /tmp/rserve.sh
    runuser -l tomcat7 /tmp/rserve.sh

    service tomcat7 stop
    echo 'JAVA_OPTS="-Xmx4096M -XX:MaxPermSize=1024M"' | sudo tee /usr/share/apache-tomcat-7/bin/setenv.sh
    wget -P /usr/share/apache-tomcat-7/webapps/ https://ci.transmartfoundation.org/browse/DEPLOY-TRAPP/latestSuccessful/artifact/shared/transmart.war/transmart.war
    patch -b -N --verbose /usr/share/apache-tomcat-7/tomcat/.grails/transmartConfig/Config.groovy #{Chef::Config[:file_cache_path]}/configGroovy.patch
    chown tomcat7 /usr/share/apache-tomcat-7/tomcat/.grails/transmartConfig/Config.groovy
    service tomcat7 start

    make -C env/ data-integration
    make -C env/ update_etl
    env > enviromentVariablesEnd

    #make -C samples/postgres load_clinical_GSE8581
    #make -C samples/postgres load_ref_annotation_GSE8581
  EOH
end

log 'rhel 6.5 Finished!'
###############################################################################
###############################################################################


recipes/rhel70.rb
log 'rhel 7.0 started'

###############################################################################
log 'Set tranSMART repository and add postgres RPM'

execute 'yum install -y http://yum.postgresql.org/9.3/redhat/rhel-7-x86_64/pgdg-redhat93-9.3-1.noarch.rpm' do
  returns [0, 1]
end

cookbook_file 'thehyve_internal.repo' do
  path '/etc/yum.repos.d/thehyve_internal.repo'
  action :create
  owner 'root'
  group 'root'
  mode 0755
end

###############################################################################
log 'Install packages'

node['unixRhel'].each do |pkg|
  package pkg do
    action :install
  end
end

node['unixNonSpecific'].each do |pkg|
  package pkg do
    action :install
  end
end

###############################################################################
log 'Configure Postgres'

execute '/usr/pgsql-9.3/bin/postgresql93-setup initdb' do
end

cookbook_file 'pg_hba.conf_trusted' do
  path '/var/lib/pgsql/9.3/data/pg_hba.conf'
  action :create
  owner 'postgres'
  group 'postgres'
  mode 0600
end

###############################################################################
log 'Install and configure Tomcat 7'

bash 'Install_tomcat7' do
  user 'root'
  cwd '/usr/share/'
  code <<-EOH
    wget -O apache-tomcat-7.tar.gz http://archive.apache.org/dist/tomcat/tomcat-7/v#{node['tomcatVersion']}/bin/apache-tomcat-#{node['tomcatVersion']}.tar.gz
    tar xzf apache-tomcat-7.tar.gz
    mv apache-tomcat-#{node['tomcatVersion']} apache-tomcat-7
    groupadd tomcat7
    useradd -g tomcat7 -d /usr/share/apache-tomcat-7/tomcat tomcat7
    chown -Rf tomcat7.tomcat7 /usr/share/apache-tomcat-7/
  EOH
end

# copy file to ec2-user to be later copied to correct location once it would not be overwritten.
cookbook_file 'Config.groovy' do
  path "#{Chef::Config[:file_cache_path]}/Config.groovy"
  action :create
  owner 'tomcat7'
  group 'tomcat7'
  mode 0644
end

cookbook_file 'tomcat7' do
  path '/etc/init.d/tomcat7'
  action :create
  owner 'tomcat7'
  group 'tomcat7'
  mode 0755
end

###############################################################################
log 'Get tranSMART from Git'

transmart_data = ENV['PWD'] + '/transmart-data'

include_recipe 'git'

git transmart_data  do
  repository 'https://github.com/transmart/transmart-data.git'
  revision 'master'
  action :checkout
end

cookbook_file 'vars' do
  path '/home/ec2-user/transmart-data/vars'
  action :create
  owner 'root'
  group 'root'
  mode 0755
end

###############################################################################
log 'Install tranSMART'

# needed for tomcat to run SOLR on right port
cookbook_file 'configGroovy.patch' do
  path "#{Chef::Config[:file_cache_path]}/configGroovy.patch"
  action :create
end

bash 'Install_tranSMART' do
  user 'root'
  cwd transmart_data
  code <<-EOH
    apachectl start
    env > enviromentVariablesStart
    chgrp -R postgres /home/ec2-user
    chmod g+rx /home/ec2-user
    export TAR_COMMAND=tar
    source vars

    systemctl enable postgresql-9.3
    systemctl start postgresql-9.3
    sudo -u postgres bash -c "source vars; PGSQL_BIN=/usr/bin/ PGDATABASE=template1 make -C ddl/postgres/GLOBAL tablespaces"
    make postgres
    bash -c "source vars; TSUSER_HOME=/usr/share/apache-tomcat-7/tomcat/ make -C config/ install"
    chown -Rf tomcat7.tomcat7 /usr/share/apache-tomcat-7/
    make -C solr/ start &
    echo 'USER=tomcat7' | sudo tee /etc/default/rserve
    bash -c 'echo "nohup /opt/R/bin/R  CMD Rserve --quiet --vanilla --RS-conf /etc/Rserve.conf&" > /tmp/rserve.sh' && chmod a+rx /tmp/rserve.sh
    runuser -l tomcat7 /tmp/rserve.sh

    service tomcat7 stop
    echo 'JAVA_OPTS="-Xmx4096M -XX:MaxPermSize=1024M"' | sudo tee /usr/share/apache-tomcat-7/bin/setenv.sh
    wget -P /usr/share/apache-tomcat-7/webapps/ https://ci.transmartfoundation.org/browse/DEPLOY-TRAPP/latestSuccessful/artifact/shared/transmart.war/transmart.war
    patch -b -N /usr/share/apache-tomcat-7/tomcat/.grails/transmartConfig/Config.groovy #{Chef::Config[:file_cache_path]}/configGroovy.patch
    chown tomcat7 /usr/share/apache-tomcat-7/tomcat/.grails/transmartConfig/Config.groovy
    service tomcat7 start

    make -C env/ data-integration
    make -C env/ update_etl
    env > enviromentVariablesEnd

    make -C samples/postgres load_clinical_GSE8581
    make -C samples/postgres load_ref_annotation_GSE8581
  EOH
end

log 'rhel Finished!'
###############################################################################
###############################################################################
